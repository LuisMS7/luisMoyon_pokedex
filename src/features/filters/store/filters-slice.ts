import { create } from "zustand";
import { Filters } from "@filters/types/filters.type";

type State = {
	pokemonName: string;
	setPokemonName: (pokemonName: string) => void;
	filters: Filters;
	filterPokemons: () => void;
};

export const useFilters = create<State>((set, get) => ({
	pokemonName: "",
	setPokemonName: (pokemonName: string) => {
		set({
			pokemonName: pokemonName.toLowerCase().trim(),
		});
	},
	filters: { pokemonName: "" },
	filterPokemons: () => {
		set({
			filters: { pokemonName: get().pokemonName },
		});
	},
}));
