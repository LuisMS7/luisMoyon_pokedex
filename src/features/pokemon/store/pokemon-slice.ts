import { create } from "zustand";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";

type State = {
	pokemons: PokemonListItemApiResponse[];
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	clearPokemons: () => void;
};

const usePokemonStore = create<State>((set, get) => ({
	pokemons: [],
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => {
		const currentPokemonsList = get().pokemons;
		const updatePokemonsList = [...currentPokemonsList, ...pokemons];
		set({
			pokemons: updatePokemonsList,
		});
	},
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => {
		set({
			pokemons: pokemons,
		});
	},
	clearPokemons: () =>
		set({
			pokemons: [],
		}),
}));

export default usePokemonStore;
