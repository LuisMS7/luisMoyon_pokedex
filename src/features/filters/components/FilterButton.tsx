import React from "react";
import { Pressable, StyleSheet, Text } from "react-native";
import { perfectSize } from "@utils/perfect-size";
import { TEXT_SUBTITLE_SIZE } from "@lib/constants";
import { useFilters } from "@filters/store/filters-slice";

const FilterButton = () => {
	const filterPokemons = useFilters((state) => state.filterPokemons);
	return (
		<Pressable style={styles.container} onPress={filterPokemons}>
			<Text style={styles.text}>Filter</Text>
		</Pressable>
	);
};

const styles = StyleSheet.create({
	container: {
		backgroundColor: "black",
		height: "100%",
		width: perfectSize(100),
		borderTopRightRadius: perfectSize(20),
		borderBottomRightRadius: perfectSize(20),
		justifyContent: "center",
		alignItems: "center",
	},
	text: {
		color: "white",
		fontSize: TEXT_SUBTITLE_SIZE,
	},
});

export default FilterButton;
