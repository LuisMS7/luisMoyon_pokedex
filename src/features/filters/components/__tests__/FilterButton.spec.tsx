import React from "react";
import { act, fireEvent, render, screen } from "@testing-library/react-native";
import FilterButton from "@filters/components/FilterButton";
import { useFilters } from "@filters/store/filters-slice";

jest.mock("@filters/store/filters-slice", () => ({
	useFilters: jest.fn(),
}));

describe("FilterButton", () => {
	it("should render correctly", () => {
		render(<FilterButton />);

		expect(screen.toJSON()).toMatchSnapshot();
	});

	it("should call filterPokemons when button is pressed", () => {
		const mockFilterPokemons = jest.fn();
		(useFilters as unknown as jest.Mock).mockImplementation((selector) =>
			selector({ filterPokemons: mockFilterPokemons }),
		);
		const { getByText } = render(<FilterButton />);

		const filterButton = getByText("Filter");

		act(() => {
			fireEvent.press(filterButton);
		});

		expect(mockFilterPokemons).toHaveBeenCalledTimes(1);
	});
});
