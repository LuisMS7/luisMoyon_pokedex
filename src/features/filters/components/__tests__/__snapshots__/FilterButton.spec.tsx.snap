// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButton should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "black",
      "borderBottomRightRadius": 29.819916046431132,
      "borderTopRightRadius": 29.819916046431132,
      "height": "100%",
      "justifyContent": "center",
      "width": 149.09958023215566,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 29.819916046431132,
      }
    }
  >
    Filter
  </Text>
</View>
`;
